### PR TITLE
Feature/percolate query

### PR DESCRIPTION
**Pull request by russcam**
Closes https://github.com/elastic/elasticsearch-net/issues/2007

@Mpdreamz, I see that you already fixed the percolate stats changes in c5b71db6472c49faa60a8f5f1a5a4c88692b055d  :+1: 

@gmarz, @Mpdreamz , What are your thoughts on `ICoreProperty`? Basically, the `percolator` type only needs name, type name and index name so these remain on `IProperty` and pushed the remaining properties up into `ICoreProperty`. 

Maybe `IAnalyzableProperty` might be better, as the properties relate to analysis?

### CODE REVIEW

File: src/Nest/CommonAbstractions/Infer/Id/IdExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Nest
+{
+    public static class IdExtensions
+    {
+		public static bool IsConditionless(this Id id)
```
Comment (gmarz) [inline src/Nest/CommonAbstractions/Infer/Id/IdExtensions.cs]: Should we make these extensions internal?  Not just this one in particular, but any `IsCondtionless()` extension (i.e. `TypeName.IsConditionless()`).

File: src/Nest/Mapping/Types/Core/Text/TextProperty.cs [hunk 2]
```diff
@@ -8,7 +8,7 @@
 namespace Nest
 {
 	[JsonObject(MemberSerialization.OptIn)]
-	public interface ITextProperty : IProperty
+	public interface ITextProperty : ICoreProperty
```
Comment (gmarz) [inline src/Nest/Mapping/Types/Core/Text/TextProperty.cs]: `ICoreProperty` has `DocValues` which isn't valid for text fields (see #2059).

**Comment by gmarz** [comment]
Few minor nitpicks but overall LGTM :+1: .  I'm +1 on `ICoreProperty`, but I think there are a few instances where it exposes properties that aren't valid for a type (e.g. `TextProperty.DocValues`).  There could be other instances as well.  We should re-evaluate all of our types as part of #2059, but I'm cool with this for alpha1.

File: src/Nest/CommonAbstractions/Infer/Id/IdExtensions.cs (hunk 1 above)
Comment (russcam) [inline src/Nest/CommonAbstractions/Infer/Id/IdExtensions.cs]: ++

File: src/Nest/Mapping/Types/Core/Text/TextProperty.cs (hunk 2 above)
Comment (russcam) [inline src/Nest/Mapping/Types/Core/Text/TextProperty.cs]: Will split DocValues out into a separate interface, and re-evaluate others as part of #2059

**Pull request merged**